lib/Dialect/SV/Transforms/HWDeadCodeElim.cpp

```diff
@@ -0,0 +1,397 @@
+  if (liveValues.insert(value).second)
+    q.push_back(value);
+
+  // If the op is within a nested region, mark the operands of all parent
+  // ops as live to capture implicit control dependencies.
+  Operation *op = value.getDefiningOp();
+  while (op && !isa<hw::HWModuleOp>(op)) {
+    for (auto operand : op->getOperands())
+      if (liveValues.insert(operand).second)
+        q.push_back(operand);
+    op = op->getParentOp();
+  }
+}
+
+void Liveness::markLive(Operation *op) {
+  for (auto result : op->getResults())
+    if (liveValues.insert(result).second)
+      q.push_back(result);
+
+  while (op && !isa<hw::HWModuleOp>(op)) {
+    for (auto operand : op->getOperands())
+      if (liveValues.insert(operand).second)
+        q.push_back(operand);
+    op = op->getParentOp();
+  }
+}
+
+void Liveness::propagate(Value result, sv::WireOp wire) {
+  for (auto *user : wire->getUsers()) {
+    if (!isa<sv::AssignOp, sv::PAssignOp, sv::BPAssignOp>(user))
+      continue;
+    if (user->getOperand(0) != wire)
+      continue;
+    markLive(user->getOperand(1));
+  }
+}
+
+void Liveness::propagate(Value result, hw::InstanceOp inst) {
+  Operation *targetOp = instanceGraph.getReferencedModule(inst);
+  if (auto module = dyn_cast<hw::HWModuleOp>(targetOp)) {
+    // If the target is a module, mark the corresponding output.
+    Block *body = module.getBodyBlock();
+    auto term = cast<hw::OutputOp>(body->getTerminator());
+    unsigned portNo = result.cast<OpResult>().getResultNumber();
+    markLive(term.getOutputs()[portNo]);
+  } else {
+    // Otherwise, over-approximate and mark all operands.
+    for (auto operand : inst->getOperands())
+      markLive(operand);
+  }
+}
+
+//===----------------------------------------------------------------------===//
+// Helper to rewrite module ports and bodies
+//===----------------------------------------------------------------------===//
+
+namespace {
+class ModuleRewriter {
+public:
+  ModuleRewriter(Liveness &liveness, hw::InstanceGraph &instanceGraph)
+      : liveness(liveness), instanceGraph(instanceGraph) {}
+
+  LogicalResult run();
+
+private:
+  LogicalResult rewrite(hw::HWModuleOp module);
+
+private:
+  Liveness &liveness;
+  hw::InstanceGraph &instanceGraph;
+
+  /// For each module that was modified, keep a bit map of the live ports.
+  DenseMap<hw::HWModuleOp, std::pair<SmallBitVector, SmallBitVector>> livePorts;
+};
+} // namespace
+
+LogicalResult ModuleRewriter::run() {
+  for (auto *instanceNode : llvm::post_order(&instanceGraph)) {
+    Operation *op = instanceNode->getModule();
+    if (!op)
+      continue;
+    auto module = dyn_cast<hw::HWModuleOp>(op);
+    if (!module)
+      continue;
+    if (failed(rewrite(module)))
+      return failure();
+  }
+  return success();
+}
+
+LogicalResult ModuleRewriter::rewrite(hw::HWModuleOp module) {
+  LLVM_DEBUG(llvm::dbgs() << module.getName() << ":\n");
+  SmallVector<unsigned> eraseInputs, eraseOutputs;
```

Comment:
Should this be conditional on the visibility of the module? I don't think we can remove ports from a `public` module.

Reply:
It is - public module ports (block args + output values) are marked as live so they are never eliminated. An assertion below also enforces that.

Reply:
Oh yeah, [here](https://github.com/llvm/circt/pull/4296/files#diff-03dfd21daeda0447d7a9e9d85669ad3df8414e8d81fb4db41de4417cafee7977R75-R83) it is! Fantastic :+1: